PR: https://github.com/odoo/odoo/pull/

From: 49c159a9060af25b49669d3b18079a95fac6229b
From: Srushti Patel
Date: 2017-04-25 09:08:14

Structural Changes: 4
Total Changes: 24

[IMP] purchase, sale_stock, stock_account: to refund purchase

When we have purchase orders and we return some products to the
supplier, we now have the possibilty to set those moves as 'to refund'
and so the quantity received is decreased.

Because the feature 'to refund' already exists on the sale orders, we
will generalize it in the module 'stock_account' which is a dependency
of both 'sale_stock' and 'purchase' which are the both module that use
the feature 'to refund'.

================================= pseudo patch: =================================

--- a/addons/stock_account/models/stock.py
+++ b/addons/stock_account/models/stock.py
@@ -176,6 +176,9 @@ class StockQuant(models.Model):
 class StockMove(models.Model):
     _inherit = "stock.move"
 
+    to_refund = fields.Boolean(string="To Refund",
+                               help='Trigger a decrease of the delivered/received quantity in the associated Sale Order/Purchase Order')
+
     @api.multi
     def action_done(self):
         self.product_price_update_before_done()
@@ -346,3 +349,24 @@ class StockMove(models.Model):
             }
             res.append((0, 0, price_diff_line))
         return res
+
+
+class StockReturnPicking(models.TransientModel):
+    _inherit = "stock.return.picking"
+
+    @api.multi
+    def _create_returns(self):
+        new_picking_id, pick_type_id = super(StockReturnPicking, self)._create_returns()
+        new_picking = self.env['stock.picking'].browse([new_picking_id])
+        for move in new_picking.move_lines:
+            return_picking_line = self.product_return_moves.filtered(lambda r: r.move_id == move.origin_returned_move_id)
+            if return_picking_line and return_picking_line.to_refund:
+                move.to_refund = True
+
+        return new_picking_id, pick_type_id
+
+
+class StockReturnPickingLine(models.TransientModel):
+    _inherit = "stock.return.picking.line"
+
+    to_refund = fields.Boolean(string="To Refund", help='Trigger a decrease of the delivered/received quantity in the associated Sale Order/Purchase Order')
